libraries/abstractions/secure_sockets/test/iot_test_udp.c

```diff
@@ -1123,6 +1199,22 @@ static void prvSOCKETS_Threadsafe_SameSocketDifferentTasks()
 
             xResult = prvCloseHelper( xSocket, &xSocketOpen );
             TEST_ASSERT_EQUAL_INT32_MESSAGE( SOCKETS_ERROR_NONE, xResult, "Socket failed to close" );
+
+            /* retry if one of the task failed. */
+            if( ( xRecvSuccess == pdTRUE ) && ( xSendSuccess == pdTRUE ) )
+            {
+                ulRetry = 0;
+                xRecvLoop++;
+            }
+            else
+            {
+                ulRetry++;
+
+                if( ulRetry > udptestMAX_RETRY )
+                {
+                    TEST_FAIL();
+                }
+            }
```

Comment:
This strikes me as convoluted.  I think it would be simpler to track total successes and total failures, where passing is having enough successes without passing a failure threshold.

Reply:
This was one of the design choice. The reason retry is used rather than finding the success rate is retry is faster and feels simpler. Taking 10% loop fail rate and 100 loop as an example.

With the retry method where number of retry (on top of the first try) is greater than 2, on average the test will perform 111.1 loops. (Out of 100 initial loops 10 loops will retry the first time, out of the 10 retries 1 loop will do another retry, and so on.) Calculating the test fail rate given 2 max retries in addition to the first try, we get 9.5%. Each incremental of max retry will reduce the test fail rate by factor of approximately 10, also each incremental of max retry does not increase the number of loop the test perform thus time is saved.

Using the success rate method, and again working with 100 loops. If we set the threshold to 90, the test will fail 54% of the time. We can reduce the test fail rate by 1) increase number of loops or 2) reduce the threshold. Increasing the number of loop will increase the number of time it take and will only converge to the theoretical number (which is 50%). If we were to reduce the threshold we can set any number between 1 and number of loops such that test will only fail say once every 1000 tests. In this case, we are reducing the number of success loop in exchange for higher chance of test passing, and if we want to mitigate and increase the number of success loop we will have to increase the number of loop it will run which will increase the amount of time the test will take. This method will require multiple variable tuning and does not perform as well as the retry method.

Reply:
There are also integration tests that only perform a few loops. May need to increasing the number of loops by 10 fold to get something statistically significant.

Reply:
Modified to threshold method